File: cairnkit/knowledge/extract_gate.py
"""Strict extraction gate (M6) — keep the knowledge base small and precise.

The archiver agent proposes candidate knowledge (semantic work it does by reading the run's
artifacts); this gate is the deterministic filter that decides which candidates are worth
storing. Borrowed from MemOS CREATE_EVAL: a candidate must be reproducible, transferable, and
carry technical depth, or it is dropped. Noise is worse than nothing.
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from datetime import date
from pathlib import Path

from cairnkit.knowledge import KNOWLEDGE_CLASSES, TYPES
from cairnkit.knowledge.model import Entry, Evidence, save_entry
from cairnkit.knowledge.schema import iter_errors

_MIN_BODY_CHARS = 80          # depth: a one-liner is not durable knowledge
CANDIDATES_FILE = "knowledge-candidates.json"


@dataclass(frozen=True)
class GateVerdict:
    accepted: bool
    reasons: tuple[str, ...]


def evaluate(candidate: dict) -> GateVerdict:
    """Decide whether a candidate knowledge dict passes the strict gate."""
    reasons: list[str] = []
    if not str(candidate.get("id", "")).strip():
        reasons.append("missing id")
    body = str(candidate.get("body", "")).strip()
    if len(body) < _MIN_BODY_CHARS:
        reasons.append("insufficient depth (body too short)")
    if not candidate.get("applicable_phases"):
        reasons.append("not transferable (no applicable_phases)")
    if candidate.get("type") not in TYPES:
        reasons.append("missing/invalid type")
    kc = candidate.get("knowledge_class", "point")
    if kc not in KNOWLEDGE_CLASSES:
        reasons.append("invalid knowledge_class")
    if not str(candidate.get("title", "")).strip():
        reasons.append("missing title")
    return GateVerdict(accepted=not reasons, reasons=tuple(reasons))


def filter_candidates(candidates: list[dict]) -> tuple[list[dict], list[dict]]:
    """Split candidates into (accepted, rejected-with-reasons)."""
    accepted: list[dict] = []
    rejected: list[dict] = []
    for c in candidates:
        verdict = evaluate(c)
        if verdict.accepted:
            accepted.append(c)
        else:
            rejected.append({"title": c.get("title", "<no title>"), "reasons": list(verdict.reasons)})
    return accepted, rejected


def _candidate_to_entry(c: dict, now: str) -> Entry:
    return Entry(
        id=str(c["id"]),
        title=str(c.get("title", "")),
        category=str(c.get("category", "")),
        domain=c.get("domain"),
        type=str(c.get("type", "")),
        guideline_polarity=c.get("guideline_polarity"),
        maturity="draft",                       # extraction always lands as draft
        knowledge_class=str(c.get("knowledge_class") or "point"),
        layer=str(c.get("layer", "L3")),
        tags=tuple(c.get("tags") or ()),
        applicable_phases=tuple(c.get("applicable_phases") or ()),
        evidence=Evidence(contributors=tuple(c.get("contributors") or ())),
        history=({"date": now, "update_type": "extract", "by": "archiver"},),
        body=str(c.get("body", "")),
    )


def _entry_path(kb_root: Path, entry: Entry) -> Path:
    if entry.category == "biz":
        return kb_root / "biz-wiki" / (entry.domain or "_") / f"{entry.id}.md"
    return kb_root / "tech-wiki" / f"{entry.id}.md"


def extract_from_run(run_dir: Path, kb_root: Path, now: str | None = None) -> dict:
    """Read the archiver's candidate file, apply the gate + schema, write accepted drafts."""
    now = now or date.today().isoformat()
    candidates_file = run_dir / CANDIDATES_FILE
    if not candidates_file.exists():
        return {"written": [], "rejected": [], "note": f"no {CANDIDATES_FILE} in {run_dir}"}
    try:
        candidates = json.loads(candidates_file.read_text(encoding="utf-8"))
    except json.JSONDecodeError as exc:
        return {"written": [], "rejected": [], "error": f"malformed {CANDIDATES_FILE}: {exc}"}
    if not isinstance(candidates, list):
        return {"written": [], "rejected": [], "error": f"{CANDIDATES_FILE} must be a JSON list"}
    accepted, rejected = filter_candidates(candidates)

    written: list[str] = []
    for c in accepted:
        entry = _candidate_to_entry(c, now)
        errs = list(iter_errors(entry))
        if errs:
            rejected.append({"title": entry.title, "reasons": [f"schema: {'; '.join(errs)}"]})
            continue
        save_entry(_entry_path(kb_root, entry), entry)
        written.append(entry.id)
    return {"written": written, "rejected": rejected}

File: src/knowledge/extract_gate.rs
//! Strict extraction gate (M6) — keep the knowledge base small and precise.

use crate::knowledge::model::{save_entry, Entry, Evidence};
use crate::knowledge::schema::iter_errors;
use crate::knowledge::{KNOWLEDGE_CLASSES, TYPES};
use chrono::Local;
use serde_json::{json, Value};
use std::path::{Path, PathBuf};

const MIN_BODY_CHARS: usize = 80;
pub const CANDIDATES_FILE: &str = "knowledge-candidates.json";

/// Decide whether a candidate dict passes the strict gate; return the rejection reasons.
pub fn evaluate(c: &Value) -> Vec<String> {
    let mut reasons = Vec::new();
    let s = |k: &str| c.get(k).and_then(|v| v.as_str()).unwrap_or("");
    if s("id").trim().is_empty() {
        reasons.push("missing id".to_string());
    }
    if s("body").trim().chars().count() < MIN_BODY_CHARS {
        reasons.push("insufficient depth (body too short)".to_string());
    }
    let phases_ok = c
        .get("applicable_phases")
        .and_then(|v| v.as_array())
        .map(|a| !a.is_empty())
        .unwrap_or(false);
    if !phases_ok {
        reasons.push("not transferable (no applicable_phases)".to_string());
    }
    if !TYPES.contains(&s("type")) {
        reasons.push("missing/invalid type".to_string());
    }
    let kc = c.get("knowledge_class").and_then(|v| v.as_str()).unwrap_or("point");
    if !KNOWLEDGE_CLASSES.contains(&kc) {
        reasons.push("invalid knowledge_class".to_string());
    }
    if s("title").trim().is_empty() {
        reasons.push("missing title".to_string());
    }
    reasons
}

fn str_field(c: &Value, k: &str, default: &str) -> String {
    c.get(k).and_then(|v| v.as_str()).unwrap_or(default).to_string()
}
fn vec_field(c: &Value, k: &str) -> Vec<String> {
    c.get(k)
        .and_then(|v| v.as_array())
        .map(|a| a.iter().filter_map(|x| x.as_str().map(String::from)).collect())
        .unwrap_or_default()
}

fn candidate_to_entry(c: &Value, now: &str) -> Entry {
    let domain = c.get("domain").and_then(|v| v.as_str()).map(String::from);
    let polarity = c.get("guideline_polarity").and_then(|v| v.as_str()).map(String::from);
    let mut hist = serde_yaml::Mapping::new();
    hist.insert("date".into(), now.into());
    hist.insert("update_type".into(), "extract".into());
    hist.insert("by".into(), "archiver".into());
    let kc = {
        let v = str_field(c, "knowledge_class", "point");
        if v.is_empty() { "point".to_string() } else { v }
    };
    Entry {
        id: str_field(c, "id", ""),
        title: str_field(c, "title", ""),
        category: str_field(c, "category", ""),
        domain,
        kind: str_field(c, "type", ""),
        guideline_polarity: polarity,
        maturity: "draft".to_string(), // extraction always lands as draft
        knowledge_class: kc,
        layer: str_field(c, "layer", "L3"),
        tags: vec_field(c, "tags"),
        applicable_phases: vec_field(c, "applicable_phases"),
        evidence: Evidence { contributors: vec_field(c, "contributors"), ..Default::default() },
        history: vec![serde_yaml::Value::Mapping(hist)],
        body: str_field(c, "body", ""),
        path: None,
    }
}

fn entry_path(kb_root: &Path, e: &Entry) -> PathBuf {
    if e.category == "biz" {
        kb_root.join("biz-wiki").join(e.domain.clone().unwrap_or_else(|| "_".to_string())).join(format!("{}.md", e.id))
    } else {
        kb_root.join("tech-wiki").join(format!("{}.md", e.id))
    }
}

pub fn extract_from_run(run_dir: &Path, kb_root: &Path) -> Value {
    let now = Local::now().format("%Y-%m-%d").to_string();
    let cf = run_dir.join(CANDIDATES_FILE);
    if !cf.exists() {
        return json!({"written": [], "rejected": [], "note": format!("no {CANDIDATES_FILE} in {}", run_dir.display())});
    }
    let text = match std::fs::read_to_string(&cf) {
        Ok(t) => t,
        Err(e) => return json!({"written": [], "rejected": [], "error": e.to_string()}),
    };
    let candidates: Value = match serde_json::from_str(&text) {
        Ok(v) => v,
        Err(e) => return json!({"written": [], "rejected": [], "error": format!("malformed {CANDIDATES_FILE}: {e}")}),
    };
    let arr = match candidates.as_array() {
        Some(a) => a,
        None => return json!({"written": [], "rejected": [], "error": format!("{CANDIDATES_FILE} must be a JSON list")}),
    };

    let mut written = Vec::new();
    let mut rejected = Vec::new();
    for c in arr {
        let reasons = evaluate(c);
        if !reasons.is_empty() {
            rejected.push(json!({"title": c.get("title").and_then(|v| v.as_str()).unwrap_or("<no title>"), "reasons": reasons}));
            continue;
        }
        let entry = candidate_to_entry(c, &now);
        let errs = iter_errors(&entry);
        if !errs.is_empty() {
            rejected.push(json!({"title": entry.title, "reasons": [format!("schema: {}", errs.join("; "))]}));
            continue;
        }
        if save_entry(&entry_path(kb_root, &entry), &entry).is_ok() {
            written.push(entry.id.clone());
        }
    }
    json!({"written": written, "rejected": rejected})
}
